src/backend/common/CommandBuffer.cpp

```diff
@@ -354,18 +452,42 @@ namespace backend {
                     }
                     break;
 
+                case Command::EndRenderPass:
+                    {
+                        iterator.NextCommand<EndRenderPassCmd>();
+                        if (currentRenderPass == nullptr) {
+                            device->HandleError("No render pass is currently active");
+                            return false;
+                        }
+                        if (currentSubpass < currentRenderPass->GetSubpassCount() - 1) {
+                            device->HandleError("Can't end a render pass before the last subpass");
+                            return false;
+                        }
+                        currentRenderPass = nullptr;
+                        currentFramebuffer = nullptr;
+                        aspects.reset(VALIDATION_ASPECT_RENDER_PASS);
+                        aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
+                    }
+                    break;
+
                 case Command::SetPipeline:
                     {
                         SetPipelineCmd* cmd = iterator.NextCommand<SetPipelineCmd>();
                         PipelineBase* pipeline = cmd->pipeline.Get();
                         PipelineLayoutBase* layout = pipeline->GetLayout();
 
                         if (pipeline->IsCompute()) {
+                            if (currentRenderPass) {
+                                device->HandleError("Can't use a compute pipeline while a render pass is active");
+                                return false;
+                            }
                             aspects.set(VALIDATION_ASPECT_COMPUTE_PIPELINE);
-                            aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
                         } else {
+                            if (!currentRenderPass) {
+                                device->HandleError("A render pass must be active when a render pipeline is set");
+                                return false;
+                            }
                             aspects.set(VALIDATION_ASPECT_RENDER_PIPELINE);
```

Comment:
nit: need to check that the pipeline's renderpass is compatible with currentRenderpass.